src/hotspot/share/gc/shared/oopStorage.cpp

```diff
@@ -880,67 +880,52 @@ bool OopStorage::should_report_num_dead() const {
 // there may be cleanup work for any storage object, based on the global
-// request state.  But that notification is deferred if the service thread
-// has run recently, and we also avoid duplicate notifications.  The service
-// thread updates the timestamp and resets the state flags on every iteration.
+// request state.  We don't want to run empty block cleanup too often in the
+// face of frequent explicit ServiceThread wakeups, hence the defer period.
 
 // Global cleanup request state.
 static volatile bool needs_cleanup_requested = false;
 
-// Flag for avoiding duplicate notifications.
-static bool needs_cleanup_triggered = false;
-
 // Time after which a notification can be made.
-static jlong cleanup_trigger_permit_time = 0;
-
-// Minimum time since last service thread check before notification is
-// permitted.  The value of 500ms was an arbitrary choice; frequent, but not
-// too frequent.
-const jlong cleanup_trigger_defer_period = 500 * NANOSECS_PER_MILLISEC;
-
-void OopStorage::trigger_cleanup_if_needed() {
-  MonitorLocker ml(Service_lock, Monitor::_no_safepoint_check_flag);
-  if (Atomic::load(&needs_cleanup_requested) &&
-      !needs_cleanup_triggered &&
-      (os::javaTimeNanos() > cleanup_trigger_permit_time)) {
-    needs_cleanup_triggered = true;
-    ml.notify_all();
-  }
-}
+static jlong cleanup_permit_time = 0;
+
+// Minimum time since last ServiceThread check before cleanup is permitted.
+// The value of 500ms was an arbitrary choice; frequent, but not too frequent.
+const jlong cleanup_defer_period = 500 * NANOSECS_PER_MILLISEC;
 
 bool OopStorage::has_cleanup_work_and_reset() {
   assert_lock_strong(Service_lock);
-  cleanup_trigger_permit_time =
-    os::javaTimeNanos() + cleanup_trigger_defer_period;
-  needs_cleanup_triggered = false;
-  // Set the request flag false and return its old value.
-  // Needs to be atomic to avoid dropping a concurrent request.
-  // Can't use Atomic::xchg, which may not support bool.
-  return Atomic::cmpxchg(&needs_cleanup_requested, true, false);
+
+  if (Atomic::load(&needs_cleanup_requested) && os::javaTimeNanos() > cleanup_permit_time) {
+    cleanup_permit_time =
+      os::javaTimeNanos() + cleanup_defer_period;
+    // Set the request flag false and return its old value.
+    // Needs to be atomic to avoid dropping a concurrent request.
+    Atomic::release_store(&needs_cleanup_requested, false);
+    return true;
+  } else {
+    return false;
+  }
 }
 
-// Record that cleanup is needed, without notifying the Service thread.
-// Used by release(), where we can't lock even Service_lock.
+// Record that cleanup is needed, without notifying the Service thread, because
+// we can't lock the Service_lock.  Used by release().
 void OopStorage::record_needs_cleanup() {
-  // Set local flag first, else service thread could wake up and miss
+  // Set local flag first, else ServiceThread could wake up and miss
   // the request.  This order may instead (rarely) unnecessarily notify.
```

Comment:
There's no longer any notification involved.  However, there is still the (rare) possibility that the ServiceThread
will uselessly run.  It might have already been doing cleanup and processed the block just added.  If no new
cleanup work gets added before the next ServiceThread cleanup time, it will attempt cleanup (because of the
flag(s) being set), and find nothing to do.  That's okay.  Or just delete the sentence about unnecessary notify.

Reply:
I deleted the sentence.  It wasn't clear what it meant anyway.  There is a sentence about notification in the comment above the function but I think that's trying to say why this doesn't do a notification like all other ServiceThread cleanups, so is still saying something that might be useful.